### PR TITLE
rocsparse CPR initial version

### PR DESCRIPTION
**Pull request by razvnane**
This PR adds support for applying CPR on AMD GPU using the rocsparse library. Similarly to the opencl implementation, the creation of AMG is done using the DUNE library.

Benchmark: NORNE
System:    CPU: AMD EPYC 7763 64-Core, 2.45GHz, GPU: AMD MI210

---

### Results ###

Units of measurement
* Total time [seconds]
* Assembly time [seconds]
* Linear solve time [seconds]
* Update time [seconds]
* Pre/post step [seconds]
* Overall Linearizations [count]
* Overall Newton Iterations [count]
* Overall Linear Iterations [count]

Case | Total Time | Assembly | Linear Solve | Update | Pre/Post | Linearisations | Newton | Linear Iterations
----- | ----------: | ----------: | -------------: | --------: | ---------: | --------------: | --------: | ---------------:
ilu0, jacobiBlocks=0, -coupled | 333 |74 | **97** | 76 | 82 | 1952 | 1582 | 22155
ilu0, jacobiBlocks=150, -coupled | 290 | 77 | **55** | 74 | 80 | 1951 | 1583 | 26336
cpr, jacobiBlocks=0, -coupled | 270 | 60 | **71** | 58 | 76 | 1571 | 1222 | 3273
cpr, jacobiBlocks=150, -coupled | 279 | 65 | **57** | 70 | 84 | 1560 | 1211 | 3557

### CODE REVIEW

File: opm/simulators/linalg/bda/opencl/openclBILU0.cpp [hunk 1]
```diff
@@ -31,6 +31,12 @@
 
 #include <sstream>
 
+#if HAVE_OPENMP
+#include <thread>
+#include <omp.h>
+extern std::shared_ptr<std::thread> copyThread;
+#endif //HAVE_OPENMP
+
```
Comment (blattms) [inline opm/simulators/linalg/bda/opencl/openclBILU0.cpp:39]: Please explain why we mix `std::thread` and OpenMP.

The former should work without the latter.

File: opm/simulators/linalg/bda/opencl/openclBILU0.cpp [hunk 2]
```diff
@@ -189,6 +195,14 @@ create_preconditioner(BlockedMatrix<Scalar>* mat, BlockedMatrix<Scalar>* jacMat)
 
     auto *matToDecompose = jacMat ? jacMat : mat;
 
+    if (jacMat) {
+#if HAVE_OPENMP
+        if (omp_get_max_threads() > 1) {
+           copyThread->join();
+        }
+#endif
```
Comment (blattms) [inline opm/simulators/linalg/bda/opencl/openclBILU0.cpp:203]: So without OpenMP we simply use no threads. If I configure OPM on a multicore machine without threads then copying will be done sequentially. Why do we want that?

File: opm/simulators/linalg/bda/opencl/openclSolverBackend.cpp [hunk 3]
```diff
@@ -73,14 +73,11 @@ openclSolverBackend(int verbosity_,
...
-                                          opencl_ilu_parallel, verbosity);
+        prec = openclPreconditioner<Scalar,block_size>::create(PreconditionerType::CPR,verbosity,opencl_ilu_parallel);
     } else if (use_isai) {
-        prec = openclPreconditioner<Scalar,block_size>::create(PreconditionerType::BISAI,
-                                          opencl_ilu_parallel, verbosity);
+        prec = openclPreconditioner<Scalar,block_size>::create(PreconditionerType::BISAI,verbosity,opencl_ilu_parallel);
     } else {
-        prec = openclPreconditioner<Scalar,block_size>::create(PreconditionerType::BILU0,
-                                          opencl_ilu_parallel, verbosity);
+        prec = openclPreconditioner<Scalar,block_size>::create(PreconditionerType::BILU0,verbosity,opencl_ilu_parallel);
```
Comment (blattms) [inline opm/simulators/linalg/bda/opencl/openclSolverBackend.cpp:80]: nitpick: Please insert spaces after commas for readability.

**Review by blattms — COMMENTED** [review]
This is surprisingly short. Was the implementation of CPR for rocSparse in  a previous PR?

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
Unfortunately there are compile errors on jenkins:
```
[236/478] Building CXX object CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o
FAILED: CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o 
/usr/lib/ccache/g++-10 -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_CHRONO_DYN_LINK -DBOOST_DATE_TIME_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_LOG_DYN_LINK -DBOOST_LOG_SETUP_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DCL_HPP_MINIMUM_OPENCL_VERSION=120 -DCL_HPP_TARGET_OPENCL_VERSION=120 -DCL_TARGET_OPENCL_VERSION=120 -DENABLE_TBB=1 -DFMT_LOCALE -DFMT_SHARED -DHAVE_CONFIG_H=1 -DMETIS_API_VERSION=5 -DSCOTCH_METIS_VERSION=5 -DVEXCL_BACKEND_OPENCL -DVEXCL_HAVE_OPENCL_HPP -D_TBB_CPP0X -I/var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators -I/var/lib/jenkins/workspace/opm-simulators-PR-builder -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/include/parmetis -I/usr/include/superlu -I/usr/include/suitesparse -I/var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/install/include -I/usr/include/cjson -I/usr/include/trilinos -I/usr/include/scotch -isystem /var/lib/jenkins/amgcl/install/include -isystem /usr/include/metis -isystem /var/lib/jenkins/damaris/install/include -isystem /usr/include/hdf5/openmpi -pipe -Wall -Wextra -Wshadow -fopenmp -pthread -O3 -DNDEBUG -mtune=native -UNDEBUG -fPIC -fopenmp -std=c++17 -MD -MT CMakeFiles/opmsimulators.dir/opm/simulators/linalg/b[236/478] Building CXX object CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o
FAILED: CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o 
/usr/lib/ccache/g++-10 -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_CHRONO_DYN_LINK -DBOOST_DATE_TIME_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_LOG_DYN_LINK -DBOOST_LOG_SETUP_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DCL_HPP_MINIMUM_OPENCL_VERSION=120 -DCL_HPP_TARGET_OPENCL_VERSION=120 -DCL_TARGET_OPENCL_VERSION=120 -DENABLE_TBB=1 -DFMT_LOCALE -DFMT_SHARED -DHAVE_CONFIG_H=1 -DMETIS_API_VERSION=5 -DSCOTCH_METIS_VERSION=5 -DVEXCL_BACKEND_OPENCL -DVEXCL_HAVE_OPENCL_HPP -D_TBB_CPP0X -I/var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators -I/var/lib/jenkins/workspace/opm-simulators-PR-builder -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/include/parmetis -I/usr/include/superlu -I/usr/include/suitesparse -I/var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/install/include -I/usr/include/cjson -I/usr/include/trilinos -I/usr/include/scotch -isystem /var/lib/jenkins/amgcl/install/include -isystem /usr/include/metis -isystem /var/lib/jenkins/damaris/install/include -isystem /usr/include/hdf5/openmpi -pipe -Wall -Wextra -Wshadow -fopenmp -pthread -O3 -DNDEBUG -mtune=native -UNDEBUG -fPIC -fopenmp -std=c++17 -MD -MT CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o -MF CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o.d -o CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o -c /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/CprCreation.cpp
In file included from /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/CprCreation.cpp:35:
/var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/Misc.hpp:4:10: fatal error: hip/hip_runtime_api.h: No such file or directory
    4 | #include <hip/hip_runtime_api.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.da/CprCreation.cpp.o -MF CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o.d -o CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o -c /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/CprCreation.cpp
In file included from /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/CprCreation.cpp:35:
/var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/Misc.hpp:4:10: fatal error: hip/hip_runtime_api.h: No such file or directory
    4 | #include <hip/hip_runtime_api.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

File: opm/simulators/linalg/bda/opencl/openclSolverBackend.cpp (hunk 3 above)
Comment (razvnane) [reply opm/simulators/linalg/bda/opencl/openclSolverBackend.cpp]: fixed

File: opm/simulators/linalg/bda/opencl/openclBILU0.cpp (hunk 1 above)
Comment (razvnane) [reply opm/simulators/linalg/bda/opencl/openclBILU0.cpp]: Yes, your are right, it should not and I fixed it now.

**Comment by razvnane** [comment]
> This is surprisingly short. Was the implementation of CPR for rocSparse in a previous PR?

I am not sure what you mean. I just ported the opencl CPR implementation that Tong Dong did and I reused the AMG creation with dune and put it in a new class CprCreation. But everything related to CPR is in this PR.

**Comment by razvnane** [comment]
> Unfortunately there are compile errors on jenkins:
> 
> ```
> [236/478] Building CXX object CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o
> FAILED: CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o 
> /usr/lib/ccache/g++-10 -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_CHRONO_DYN_LINK -DBOOST_DATE_TIME_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_LOG_DYN_LINK -DBOOST_LOG_SETUP_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DCL_HPP_MINIMUM_OPENCL_VERSION=120 -DCL_HPP_TARGET_OPENCL_VERSION=120 -DCL_TARGET_OPENCL_VERSION=120 -DENABLE_TBB=1 -DFMT_LOCALE -DFMT_SHARED -DHAVE_CONFIG_H=1 -DMETIS_API_VERSION=5 -DSCOTCH_METIS_VERSION=5 -DVEXCL_BACKEND_OPENCL -DVEXCL_HAVE_OPENCL_HPP -D_TBB_CPP0X -I/var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators -I/var/lib/jenkins/workspace/opm-simulators-PR-builder -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/include/parmetis -I/usr/include/superlu -I/usr/include/suitesparse -I/var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/install/include -I/usr/include/cjson -I/usr/include/trilinos -I/usr/include/scotch -isystem /var/lib/jenkins/amgcl/install/include -isystem /usr/include/metis -isystem /var/lib/jenkins/damaris/install/include -isystem /usr/include/hdf5/openmpi -pipe -Wall -Wextra -Wshadow -fopenmp -pthread -O3 -DNDEBUG -mtune=native -UNDEBUG -fPIC -fopenmp -std=c++17 -MD -MT CMakeFiles/opmsimulators.dir/opm/simulators/linalg/b[236/478] Building CXX object CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o
> FAILED: CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o 
> /usr/lib/ccache/g++-10 -DBOOST_ALL_NO_LIB -DBOOST_ATOMIC_DYN_LINK -DBOOST_CHRONO_DYN_LINK -DBOOST_DATE_TIME_DYN_LINK -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_LOG_DYN_LINK -DBOOST_LOG_SETUP_DYN_LINK -DBOOST_REGEX_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DCL_HPP_MINIMUM_OPENCL_VERSION=120 -DCL_HPP_TARGET_OPENCL_VERSION=120 -DCL_TARGET_OPENCL_VERSION=120 -DENABLE_TBB=1 -DFMT_LOCALE -DFMT_SHARED -DHAVE_CONFIG_H=1 -DMETIS_API_VERSION=5 -DSCOTCH_METIS_VERSION=5 -DVEXCL_BACKEND_OPENCL -DVEXCL_HAVE_OPENCL_HPP -D_TBB_CPP0X -I/var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/build-opm-simulators -I/var/lib/jenkins/workspace/opm-simulators-PR-builder -I/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -I/usr/lib/x86_64-linux-gnu/openmpi/include -I/usr/include/parmetis -I/usr/include/superlu -I/usr/include/suitesparse -I/var/lib/jenkins/workspace/opm-simulators-PR-builder/mpi/install/include -I/usr/include/cjson -I/usr/include/trilinos -I/usr/include/scotch -isystem /var/lib/jenkins/amgcl/install/include -isystem /usr/include/metis -isystem /var/lib/jenkins/damaris/install/include -isystem /usr/include/hdf5/openmpi -pipe -Wall -Wextra -Wshadow -fopenmp -pthread -O3 -DNDEBUG -mtune=native -UNDEBUG -fPIC -fopenmp -std=c++17 -MD -MT CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o -MF CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o.d -o CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o -c /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/CprCreation.cpp
> In file included from /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/CprCreation.cpp:35:
> /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/Misc.hpp:4:10: fatal error: hip/hip_runtime_api.h: No such file or directory
>     4 | #include <hip/hip_runtime_api.h>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.da/CprCreation.cpp.o -MF CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o.d -o CMakeFiles/opmsimulators.dir/opm/simulators/linalg/bda/CprCreation.cpp.o -c /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/CprCreation.cpp
> In file included from /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/CprCreation.cpp:35:
> /var/lib/jenkins/workspace/opm-simulators-PR-builder/opm/simulators/linalg/bda/Misc.hpp:4:10: fatal error: hip/hip_runtime_api.h: No such file or directory
>     4 | #include <hip/hip_runtime_api.h>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> ```

This should also be fixed now. Please run jenkins again.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
Still compilation failures:
```
tests/test_solvetransposed3x3.cpp:34:10: fatal error: opm/simulators/linalg/bda/opencl/CPR.hpp: No such file or directory
   34 | #include <opm/simulators/linalg/bda/opencl/CPR.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[323/482] Building CXX object CMakeFiles/test_csrToCscOffsetMap.dir/tests/test_csrToCscOffsetMap.cpp.o
/test_csrToCscOffsetMap.cpp:12:10: fatal error: opm/simulators/linalg/bda/opencl/BISAI.hpp: No such file or directory
   12 | #include <opm/simulators/linalg/bda/opencl/BISAI.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Note, that there were some hickups when moving the build servers to Debian 12. Unfortunately, I cannot find @akva2's message about that.

**Comment by razvnane** [comment]
Ah yes, I forgot to update the tests after renaming the files. Should be fixed now. Please rerun jenkins.

**Comment by bska** [comment]
jenkins build this serial rocm please

**Comment by razvnane** [comment]
one more try please

**Comment by blattms** [comment]
jenkins build this serial rocm please

**Comment by bska** [comment]
For what it's worth, this  PR adds a lot of new warnings as demonstrated by [the Jenkins post-commit build](https://ci.opm-project.org/job/opm-simulators/2276/gcc/).  I would really appreciate it if we could address the ones attributed to files within the `opm/bda` directory in follow-up work.

**Pull request merged**